### ISSUE TITLE
Async way to set animation source

### ISSUE THREAD

**Issue by bitapparat** [XamlAnimatedGif/WpfAnimatedGif#29]
This library is a true lifesaver. But i'm having one issue: 
Loading a large GIF ([example](http://68.media.tumblr.com/715a020fc34bc4e8e2d4815fd01c9b3f/tumblr_nwomzjGE681qghzg0o1_500.gif)) freezes the entire UI thread for multiple seconds. I tried setting the image as invisible until the AnimationLoaded event triggers, but that doesn't help either.

The best solution to avoid such problems would probably be to have an async method to set the animated source (SetAnimatedSourceAsync). The method would do all the animation initialization beforehand in an async manner and set the Image Source when the animation loading is completed. With async/await a user would just call and await the method. No hangs and no need to deal with events like AnimationLoaded either.

**Comment by thomaslevesque** [comment]
Hi @bitapparat,

I just tried this image in the demo app (both WPF and WinRT), and didn't notice any UI freeze (although there was a rather long delay before the first frame appeared). How are you loading the image? What kind of app is it (WPF, WinRT, Windows Phone?)

**Comment by bitapparat** [comment]
It's a WPF application with the current WpfAnimatedGif NuGet package.
The stripped down code is just this:
`         var img = new BitmapImage();
         img.BeginInit();
         img.UriSource = new Uri("http://68.media.tumblr.com/715a020fc34bc4e8e2d4815fd01c9b3f/tumblr_nwomzjGE681qghzg0o1_500.gif");
         img.EndInit();
         WpfAnimatedGif.ImageBehavior.SetAnimatedSource(CurrentImage, img);
`

**Comment by thomaslevesque** [comment]
Oh, sorry. I was confusing with my other project, [XamlAnimatedGif](https://github.com/XamlAnimatedGif/XamlAnimatedGif).

OK, the way you're loading the image, it isn't actually downloaded yet after `EndInit()`. To force it to download immediately, you need to set `CacheOption` to `OnLoad`:

```
    img.BeginInit();
    img.CacheOption = BitmapCacheOption.OnLoad;
    img.UriSource = new Uri("http://68.media.tumblr.com/715a020fc34bc4e8e2d4815fd01c9b3f/tumblr_nwomzjGE681qghzg0o1_500.gif");
    img.EndInit();
```

**Comment by bitapparat** [comment]
I'm waiting on the DownloadCompleted event of the image now before calling SetAnimatedSource. I'm still getting an UI freeze of about a second. The freeze does not happen when setting Image.Source without animation instead, so it's not the download blocking the thread. Code:

         var img = new BitmapImage();
         img.BeginInit();
         img.CacheOption = BitmapCacheOption.OnLoad;
         img.DownloadCompleted += (s, args) => {
            TestImage.Dispatcher.BeginInvoke(new Action(() => WpfAnimatedGif.ImageBehavior.SetAnimatedSource(TestImage, img)));
         };
         img.UriSource = new Uri("http://68.media.tumblr.com/715a020fc34bc4e8e2d4815fd01c9b3f/tumblr_nwomzjGE681qghzg0o1_500.gif");
         img.EndInit();

**Comment by thomaslevesque** [comment]
To honest, I don't really maintain this library anymore... Most users have switched to XamlAnimatedGif by now. Is there a reason for you to keep using WpfAnimatedGif?

**Comment by bitapparat** [comment]
I tried XamlAnimatedGif, which fixed the UI freeze, but had another issue where GIF playback seemed slower than it should and started stuttering quite badly when another animated GIF was loading for another Image element. I can open in issue with a code sample on the other project, if you're interested. It's fine if not. My use case isn't worth a very thorough investigation and i am okay living with the short UI freeze of WpfAnimatedGif at this point.

**Comment by thomaslevesque** [comment]
Yeah, I know, neither project is perfect... WpfAnimatedGif uses an awful lot of RAM, XamlAnimatedGif has poor performance with large GIFs.

> I can open in issue with a code sample on the other project, if you're interested.

Thanks, but there are already quite a few issues about the performance of XamlAnimatedGif ;)

**Comment by thomaslevesque** [comment]
Closing due to inactivity
